src/Umbraco.Web/PropertyEditors/ValueConverters/MultiUrlPickerValueConverter.cs

```diff
@@ -15,77 +14,102 @@ namespace Umbraco.Web.PropertyEditors.ValueConverters
                 foreach (var dto in dtos)
                 {
+                    var name = dto.Name;
                     var type = LinkType.External;
                     var url = dto.Url;
+                    IPublishedContent content = null;
 
                     if (dto.Udi != null)
                     {
-                        type = dto.Udi.EntityType == Core.Constants.UdiEntityType.Media
-                            ? LinkType.Media
-                            : LinkType.Content;
-
-                        var content = type == LinkType.Media ?
-                             _publishedSnapshotAccessor.PublishedSnapshot.Media.GetById(preview, dto.Udi.Guid) :
-                             _publishedSnapshotAccessor.PublishedSnapshot.Content.GetById(preview, dto.Udi.Guid);
+                        switch (dto.Udi.EntityType)
+                        {
+                            case Core.Constants.UdiEntityType.Document:
+                                type = LinkType.Content;
+                                content = _publishedSnapshotAccessor.PublishedSnapshot.Content.GetById(preview, dto.Udi.Guid);
+                                break;
+                            case Core.Constants.UdiEntityType.Media:
+                                type = LinkType.Media;
+                                content = _publishedSnapshotAccessor.PublishedSnapshot.Media.GetById(preview, dto.Udi.Guid);
+                                break;
+                            default:
+                                // Skip unsupported entity types
+                                continue;
+                        }
 
-                        if (content == null || content.ContentType.ItemType == PublishedItemType.Element)
+                        if (content == null)
                         {
+                            // Skip unavailable content
                             continue;
                         }
-                        url = content.Url();
-                    }
 
-                    links.Add(
-                        new Link
+                        if (string.IsNullOrEmpty(name))
                         {
-                            Name = dto.Name,
-                            Target = dto.Target,
-                            Type = type,
-                            Udi = dto.Udi,
-                            Url = url + dto.QueryString,
+                            name = content.Name;
                         }
-                    );
+
+                        url = content.Url();
+                    }
+
+                    links.Add(new Link()
+                    {
+                        Name = name,
+                        Target = dto.Target,
+                        Type = type,
+                        Udi = dto.Udi,
+                        Content = content,
+                        Url = url + dto.QueryString
+                    });
+
+                    if (maxNumber > 0 && links.Count == maxNumber)
```

Comment:
This currently ensures that no more links are processed then needed. However, this behavior is a little bit strange, as lowering the `maxNumber` will truncate the list of already selected, valid links. Lowering the `maxChars` on a Textbox or `maxItems` on Nested Content won't do this. Yes, it should validate the maximum number when publishing, but let's keep the behavior of all default Umbraco PVCs the same.

So we only have to short-circuit when we're in single item mode:
```suggestion
                    if (maxNumber == 1)
```